core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -248,6 +259,23 @@ public MetadataManager(
         verifyTypes();
 
         functionDecoder = new ResolvedFunctionDecoder(this::getType);
+
+        operatorCache = CacheBuilder.newBuilder()
+                .maximumSize(1000)
+                .build(CacheLoader.from(entry -> {
+                    String name = mangleOperatorName(entry.getKey());
+                    return resolveFunction(QualifiedName.of(name), fromTypes(entry.getValue()));
+                }));
+
+        coercionCache = CacheBuilder.newBuilder()
+                .maximumSize(1000)
+                .build(CacheLoader.from(cell -> {
+                    String name = mangleOperatorName(cell.getRowKey());
+                    Type fromType = cell.getColumnKey();
+                    Type toType = cell.getValue();
+                    Signature signature = new Signature(name, toType.getTypeSignature(), ImmutableList.of(fromType.getTypeSignature()));
+                    return resolve(functionResolver.resolveCoercion(functions.get(QualifiedName.of(name)), signature));
```

Comment:
You can just cache `io.trino.metadata.MetadataManager#resolve(io.trino.metadata.FunctionBinding, io.trino.metadata.FunctionDependencyDeclaration)` as all other functions delegate to it eventually.

Reply:
Then FunctionResolver.resolveCoercion() is not cached. 
From what I observed, FunctionResolver.resolveFunction() and FunctionResolver.resolveCoercion() are the slow part. 
I considered adding cache in FunctionResolver, but some of the argument types are not suitable to be a part of cache key, such as TypeSignatureProvider.